rllib/algorithms/ppo/tests/test_ppo_rl_module.py

```diff
@@ -0,0 +1,242 @@
+            input_dim=obs_dim,
+            hidden_layers=[32],
+            activation="ReLU",
+            output_dim=32,
+        )
+        pi_encoder_config = IdentityConfig(output_dim=32)
+        vf_encoder_config = IdentityConfig(output_dim=32)
+    else:
+        shared_encoder_config = IdentityConfig(output_dim=obs_dim)
+        if lstm:
+            pi_encoder_config = LSTMConfig(
+                input_dim=obs_dim,
+                hidden_dim=32,
+                batch_first=True,
+                output_dim=32,
+                num_layers=1,
+            )
+        else:
+            pi_encoder_config = FCConfig(
+                input_dim=obs_dim,
+                output_dim=32,
+                hidden_layers=[32],
+                activation="ReLU",
+            )
+        vf_encoder_config = FCConfig(
+            input_dim=obs_dim,
+            output_dim=32,
+            hidden_layers=[32],
+            activation="ReLU",
+        )
+
+    pi_config = FCConfig()
+    vf_config = FCConfig()
+
+    if isinstance(env.action_space, gym.spaces.Discrete):
+        pi_config.output_dim = env.action_space.n
+    else:
+        pi_config.output_dim = env.action_space.shape[0] * 2
+
+    return PPOModuleConfig(
+        observation_space=env.observation_space,
+        action_space=env.action_space,
+        shared_encoder_config=shared_encoder_config,
+        pi_encoder_config=pi_encoder_config,
+        vf_encoder_config=vf_encoder_config,
+        pi_config=pi_config,
+        vf_config=vf_config,
+        shared_encoder=shared_encoder,
+    )
+
+
+class TestPPO(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        ray.init()
+
+    @classmethod
+    def tearDownClass(cls):
+        ray.shutdown()
+
+    def test_rollouts(self):
+        # TODO: Add BreakoutNoFrameskip-v4 to cover a 3D obs space
+        for env_name in ["CartPole-v1", "Pendulum-v1"]:
+            for fwd_fn in ["forward_exploration", "forward_inference"]:
+                for shared_encoder in [False, True]:
+                    for lstm in [True, False]:
+                        if lstm and shared_encoder:
+                            # Not yet implemented
+                            # TODO (Artur): Implement
+                            continue
+                        print(
+                            f"[ENV={env_name}] | [FWD={fwd_fn}] | [SHARED="
+                            f"{shared_encoder}] | LSTM={lstm}"
+                        )
+                        env = gym.make(env_name)
+
+                        config = get_expected_model_config(env, lstm, shared_encoder)
+                        module = PPOTorchRLModule(config)
+
+                        obs = env.reset()
+
+                        batch = {
+                            SampleBatch.OBS: convert_to_torch_tensor(obs)[None],
+                        }
+
+                        if lstm:
+                            state_in = module.pi_encoder.get_inital_state()
```

Comment:
`module.get_initial_state()`?

Reply:
module.get_initial_state() seems to not have settled yet and only returns a state `if isinstance(self.config.shared_encoder_config, LSTMConfig)`. We can test here without that having settled there for now so let's leave it for the moment.

Reply:
Dude the point of pushing for this PR is also fix the shortcomings of what's inside ppo_torch_rl_module. Please fix them if you think they are not covering some cases. But in the test, you should make it independent of implementation details. That means you should use `module.get_intial_state()` to be agonistic to the fact the module owns a pi_encoder or shared_encoder.

Reply:
Since you are the designer of the RLModule, what do you think get_initial_state should look like? Are you sure you don't simply want to do that on another PR? According to your comment in that method it looks like you have thought about this already.

Reply:
Can you explain why this code is problematic:
```
    def get_initial_state(self) -> NestedDict:
        if isinstance(self.config.shared_encoder_config, LSTMConfig):
            # TODO (Kourosh): How does this work in RLlib today?
            if isinstance(self.shared_encoder, LSTMEncoder):
                return self.shared_encoder.get_inital_state()
            else:
                return self.pi_encoder.get_inital_state()
        return {}
```

Reply:
Because it ends up never returning the initial state of pi_encoder if I'm not mistaken.

Reply:
is you shared_encoder an LSTMEncoder? if so why is pi_encoder an LSTM that you want to reach that line?

Reply:
I don't get why you don't reach `return self.pi_encoder.get_inital_state()` line

Reply:
I'm saying, what if shared encoder is for example an FC net and pi encoder is LSTM?

Reply:
Then `isinstance(self.shared_encoder, LSTMEncoder)` would return False?

Reply:
Then this method will return {}

Reply:
`if isinstance(self.config.shared_encoder_config, LSTMConfig)` will also return False

Reply:
ok then do something like this instead? you get my point? change this method's logic to something that works and use `module.get_intial_state()` not `module.xxx.get_initial_state()`.

```
def get_initial_state(self) -> NestedDict:
    if isinstance(self.shared_encoder, LSTMEncoder):
        return self.shared_encoder.get_inital_state()
    elif isinstance(self.pi_encoder, LSTMEncoder):
        return self.pi_encoder.get_inital_state()
    else:
        return {}

```

Reply:
Yeah I got that but what happens if they are both recurrent?

Reply:
What happens if you use a recurrent vf encoder? What if all of these are current?

Reply:
That's why I said this API has likely not settled and I don't want to depend on it in this test for now.

Reply:
They should not be both recurrent?? That's the whole point of encoder + pi and vf head abstraction. Also I think regardless of share=True/False, when we have lstm=True we should share an LSTM encoder. Otherwise there should be different get_states for vf and pi. There should be one lstm latent encoder (that provides history). Then we can give enough capacity to the pi and vf neural nets to apply more complicated non-linear transformation on the latent if we really have to.

Reply:
The api will gradually get settled, with small prs like this that move the requirements to their right place.

Reply:
For example from this exercise you should now have more info about the requirements of the catalog encoder/trunk generator. Don't sweep these information under the rug.

Reply:
That's fair. Let me go with your suggested update then:

` def get_initial_state(self) -> NestedDict:
        if isinstance(self.config.shared_encoder_config, LSTMConfig):
            # TODO (Kourosh): How does this work in RLlib today?
            if isinstance(self.shared_encoder, LSTMEncoder):
                return self.shared_encoder.get_inital_state()
            else:
                return self.pi_encoder.get_inital_state()
        return {}
`